src/Microsoft.Azure.NotificationHubs/NotificationHubClient.cs

```diff
@@ -2186,7 +2156,26 @@ private void SetUserAgent()
                 xmlReader.ReadToDescendant("content");
                 xmlReader.ReadStartElement();
 
-                return (TEntity)_entitySerializer.Deserialize(xmlReader, xmlReader.Name);
+                var entity = _entitySerializer.Deserialize(xmlReader, xmlReader.Name);
+
+                if (typeof(GcmRegistrationDescription).IsAssignableFrom(typeof(TEntity)))
+                {
+                    return (TEntity)entity;
+                }
+
+                if (entity is GcmTemplateRegistrationDescription gcmTemplateRegistrationDescription)
+                {
+                    var fcmTemplateRegistrationDescription = new FcmTemplateRegistrationDescription(gcmTemplateRegistrationDescription);
+                    return (fcmTemplateRegistrationDescription as TEntity);
+                }
+
+                if (entity is GcmRegistrationDescription gcmRegistrationDescription)
+                {
+                    var fcmRegistrationDescription = new FcmRegistrationDescription(gcmRegistrationDescription);
+                    return (fcmRegistrationDescription as TEntity);
```

Comment:
Is there a reason to use `as` casting versus a direct cast?  Meaning could this be `null` ever?

Reply:
@mpodwysocki  returned value could not be `null`